ingestion-beam/src/main/java/com/mozilla/telemetry/transforms/CompressPayload.java

```diff
@@ -24,38 +27,59 @@
     extends PTransform<PCollection<? extends PubsubMessage>, PCollection<PubsubMessage>> {
 
   public static CompressPayload of(ValueProvider<Compression> compression) {
-    return new CompressPayload(compression);
+    return new CompressPayload(compression, Integer.MAX_VALUE);
+  }
+
+  public CompressPayload withMaxCompressedBytes(int maxBytes) {
+    return new CompressPayload(this.compression, maxBytes);
   }
 
   @Override
   public PCollection<PubsubMessage> expand(PCollection<? extends PubsubMessage> input) {
-    return input.apply(MapElements.into(TypeDescriptor.of(PubsubMessage.class))
-        .via(message -> compress(message, compression.get())));
+    return input
+        .apply(MapElements.into(TypeDescriptor.of(PubsubMessage.class)).via(this::compress));
   }
 
   ////////
 
-  private final ValueProvider<Compression> compression;
+  final ValueProvider<Compression> compression;
+  private final int maxCompressedBytes;
 
-  private CompressPayload(ValueProvider<Compression> compression) {
+  private final Counter truncationCounter = Metrics.counter(CompressPayload.class,
+      "truncated_payload");
+
+  private CompressPayload(ValueProvider<Compression> compression, int maxCompressedBytes) {
     this.compression = compression;
+    this.maxCompressedBytes = maxCompressedBytes;
+  }
+
+  @VisibleForTesting
+  PubsubMessage compress(PubsubMessage message) {
+    message = PubsubConstraints.ensureNonNull(message);
+    byte[] compressedBytes = compress(message.getPayload(), compression.get());
+    if (compressedBytes.length > maxCompressedBytes) {
+      byte[] truncated = Arrays.copyOfRange(message.getPayload(), 0, maxCompressedBytes);
+      truncationCounter.inc();
+      return compress(new PubsubMessage(truncated, message.getAttributeMap()));
```

Comment:
why are we doing this instead of?
```suggestion
      return new PubsubMessage(compress(truncated, compression.get()), message.getAttributeMap());
```